src/js/components/devices/device-details/devicetwin.js

```diff
@@ -0,0 +1,250 @@
+    lineNumbersMinChars: 3,
+    quickSuggestions: false,
+    renderOverviewRuler: false,
+    scrollBeyondLastLine: false,
+    readOnly: true
+  }
+};
+
+const externalProvider = EXTERNAL_PROVIDER.azure;
+const indentation = 4; // number of spaces, tab based indentation won't show in the editor, but be converted to 4 spaces
+
+export const DeviceTwin = ({ device, setDeviceTwin }) => {
+  const [configured, setConfigured] = useState('');
+  const [diffCount, setDiffCount] = useState(0);
+  const [isEditing, setIsEditing] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [open, setOpen] = useState(false);
+  const [reported, setReported] = useState('');
+  const [updated, setUpdated] = useState('');
+  const [isSync, setIsSync] = useState(true);
+  const editorRef = useRef(null);
+
+  const { [externalProvider.provider]: deviceTwin = {} } = device.twinsByProvider ?? {};
+  const { desired: configuredTwin, reported: reportedTwin, updated_ts: updateTime = device.created_ts } = deviceTwin;
+
+  useEffect(() => {
+    const textContent = JSON.stringify(configuredTwin, undefined, indentation) ?? '';
+    setConfigured(textContent);
+    setUpdated(textContent);
+    setReported(JSON.stringify(reportedTwin, undefined, indentation) ?? '');
+    setIsEditing;
+  }, [open]);
+
+  useEffect(() => {
+    setIsSync(deepCompare(reported, configured));
+  }, [configured, reported]);
+
+  const handleEditorDidMount = (editor, monaco) => {
+    editorRef.current = { editor, monaco, modifiedEditor: editor };
+  };
+
+  const handleDiffEditorDidMount = (editor, monaco) => {
+    const modifiedEditor = editor.getModifiedEditor();
+    modifiedEditor.onDidChangeModelContent(() => setUpdated(modifiedEditor.getValue()));
+    editor.onDidUpdateDiff(onDidUpdateDiff);
+    editorRef.current = { editor, monaco, modifiedEditor };
+  };
+
+  const onDidUpdateDiff = () => {
+    const changes = editorRef.current.editor.getLineChanges();
+    setDiffCount(changes.length);
+  };
+
+  const onApplyClick = () => {
+    let update = {};
+    try {
+      update = JSON.parse(updated);
+    } catch (error) {
+      setErrorMessage('There was an error parsing the device twin changes, please ensure that it is valid JSON.');
+      return;
+    }
+    editorRef.current.modifiedEditor.getAction('editor.action.formatDocument').run();
+    setUpdated(JSON.stringify(update, undefined, 4));
+    setErrorMessage('');
+    setDeviceTwin(device.id, externalProvider.provider, update).then(() => {
+      setIsEditing(false);
+    });
+  };
+
+  const onCancelClick = () => {
+    const textContent = JSON.stringify(configuredTwin, undefined, indentation) ?? '';
+    setUpdated(textContent);
+    editorRef.current.modifiedEditor.getModel().setValue(textContent);
+    setIsEditing(false);
+  };
+
+  const onEditClick = () => setIsEditing(true);
+
+  return (
+    <DeviceDataCollapse
+      header={
+        !open &&
+        (twinError ? (
```

Comment:
This is throwing a "not defined" error now

Reply:
argh right... right now I'm adding the refresh button and making the external integration setup depend on the released version - once I've pushed this I'll ping you 🙂

Reply:
@michaelatmender now the refresh button should also work 👌
A note on testing: From now on you have to set `INTEGRATION_VERSION: 3.2.0` for the integration settings to show. I thought this might be the easiest way to decide if we can show it, to not require an additional flag when running the current GUI on Hosted Mender...